### PR TITLE
Cmr 4300

### PR DESCRIPTION
**Pull request by daniel-zamora**
This PR implements the search for deleted granules, to support CMR harvesting.

*NOTE: Still need to update api-docs

### CODE REVIEW

File: search-app/src/cmr/search/services/parameters/parameter_validation.clj [hunk 1]
```diff
@@ -789,3 +798,62 @@
                         deleted-colls-revision-date-validation])]
     (when (seq errors)
       (errors/throw-service-errors :bad-request errors))))
+
+(defn- deleted-grans-result-format-validation
+  "Validates that the only result format support by deleted granules search is :xml"
```
Comment (laurenfrederick) [inline search-app/src/cmr/search/services/parameters/parameter_validation.clj]: :xml should be :json

File: search-app/src/cmr/search/services/parameters/parameter_validation.clj [hunk 2]
```diff
@@ -789,3 +798,62 @@
...
+        (validate-deleted-grans-revision-date-str (first revision-date)))
+      (validate-deleted-grans-revision-date-str revision-date))))
+
+(defn- deleted-grans-revision-date-range-validation
+  "Validates that deleted time can only have a start date for deleted collections search"
+  [params]
+  (when-let [revision-date (or (:revision-date params)
+                               (:revision_date params))]
+    (when (t/before? (dt-parser/parse-datetime revision-date) (t/minus- (t/now) (t/days 365)))
+      [(format "Revision date must be withing one year of today.")])))
```
Comment (laurenfrederick) [inline search-app/src/cmr/search/services/parameters/parameter_validation.clj]: Change withing to within

File: search-app/src/cmr/search/services/query_service.clj [hunk 3]
```diff
@@ -427,6 +430,45 @@
...
      :result-format result-format}))
 
+(defn get-deleted-granules
+  "Executes elasticsearch searches to find collections that are deleted from a given revision-date.
+   This only finds granules that are truly deleted and not searchable.
+   granules that are deleted, then later ingested again are not included in the result."
+
+  [context params]
+  (pv/validate-deleted-granules-params params)
+  (let [start-time (System/currentTimeMillis)
```
Comment (laurenfrederick) [inline search-app/src/cmr/search/services/query_service.clj]: Could move some of this into a separate function, like creating the filters.

File: search-app/src/cmr/search/services/parameters/parameter_validation.clj [hunk 4]
```diff
@@ -748,6 +748,15 @@
     :revision_date
     :result-format})
 
+(def ^:private valid-deleted-granules-search-params
+  "Valid parameters for deleted granules search"
+  #{:revision-date
+    :provider
+    :parent-collection-id
+    :revision_date
```
Comment (chris-durbin) [inline search-app/src/cmr/search/services/parameters/parameter_validation.clj]: These should all be kebab-case (no underscores). Processing params should convert all underscores in query-params to dashes.

So removing :parent_collection_id and :revision_date should work without problem.

File: search-app/src/cmr/search/services/parameters/parameter_validation.clj [hunk 5]
```diff
@@ -748,6 +748,15 @@
...
+    :parent_collection_id
+    :result-format})
```
Comment (chris-durbin) [inline search-app/src/cmr/search/services/parameters/parameter_validation.clj]: We only support JSON right now and for other endpoints we support the result-format through a URL extension or header, not a query parameter. I'm wondering why this is in the list?

File: search-app/src/cmr/search/services/parameters/parameter_validation.clj [hunk 6]
```diff
@@ -789,3 +798,62 @@
...
+(defn- validate-deleted-grans-revision-date-str
+  [revision-date]
+  (when (.contains revision-date ",")
+    [(format (str "Invalid format for revision date, only a starting date is allowed "
+                  "for deleted granules search, but was [%s]") revision-date)]))
+
+(defn- deleted-grans-revision-date-validation
+  "Validates that deleted time can only have a start date for deleted collections search"
+  [params]
+  (when-let [revision-date (or (:revision-date params)
```
Comment (chris-durbin) [inline search-app/src/cmr/search/services/parameters/parameter_validation.clj]: Similar comment: params should be processed so that we only need to look for the dash case.

File: search-app/src/cmr/search/services/parameters/parameter_validation.clj [hunk 7]
```diff
@@ -789,3 +798,62 @@
...
+(defn- deleted-grans-revision-required-validation
+  "Validates that deleted time can only have a start date for deleted collections search"
+  [params]
+  (let [revision-date (or (:revision-date params)
+                          (:revision_date params))]
+    (when (not (seq revision-date))
+      [(format "One revision date is required for deleted granules search.")])))
+
+(defn validate-deleted-granules-params
+  "Validates the query parameters passed in with deleted collections search.
```
Comment (chris-durbin) [inline search-app/src/cmr/search/services/parameters/parameter_validation.clj]: granules rather than collections

File: search-app/src/cmr/search/services/parameters/parameter_validation.clj [hunk 8]
```diff
@@ -789,3 +798,62 @@
...
+   Throws exceptions to send to the user if a validation fails."
+  [params]
+  (let [errors (mapcat #(% params)
+                       [unrecognized-deleted-grans-params-validation
```
Comment (chris-durbin) [inline search-app/src/cmr/search/services/parameters/parameter_validation.clj]: Were there any validations from deleted-collections that could be reused? I was just thinking any of the date parsing, single value validations should have been identical to the collection ones.

File: search-app/src/cmr/search/services/query_service.clj [hunk 9]
```diff
@@ -16,6 +16,8 @@
    [clj-time.format :as time-format]
    [clojure.set :as set]
    [clojure.string :as string]
+   [clojurewerkz.elastisch.query :as esq]
```
Comment (chris-durbin) [inline search-app/src/cmr/search/services/query_service.clj]: With our existing design we do not perform any direct Elasticsearch queries in this namespace.

File: search-app/src/cmr/search/services/query_service.clj [hunk 10]
```diff
@@ -427,6 +430,45 @@
...
+(defn get-deleted-granules
+  "Executes elasticsearch searches to find collections that are deleted from a given revision-date.
+   This only finds granules that are truly deleted and not searchable.
+   granules that are deleted, then later ingested again are not included in the result."
+
+  [context params]
+  (pv/validate-deleted-granules-params params)
+  (let [start-time (System/currentTimeMillis)
+        {:keys [parent-collection-id provider revision-date result-format]} (common-params/sanitize-params params)
+        filters [{:range {:revision-date {:gte revision-date}}}]
```
Comment (chris-durbin) [inline search-app/src/cmr/search/services/query_service.clj]: You should be able to use existing code to help build up a query model and then call execute query so it follows through most of the same paths in the code. I'm wondering if the deleted-collections endpoint might be a good example of how we perform elasticsearch queries.

File: system-int-test/test/cmr/system_int_test/ingest/deleted_granules_index_test.clj [hunk 11]
```diff
@@ -53,3 +71,110 @@
...
+    (ingest/delete-concept granule4-prov2)
+    (index/wait-until-indexed)
+
+    (testing "Search for all"
+      (are3 [params granules]
+        (is (= (set (map :concept-id granules))
+               (set (find-deleted-granules params))))
+
+        "after tomorrow"
+        {:revision-date (t/plus- (t/now) (t/days 1))}
```
Comment (chris-durbin) [inline system-int-test/test/cmr/system_int_test/ingest/deleted_granules_index_test.clj]: I'm not sure what others think about this, but I always want to make sure we verify our query parameters on the API support underscores. I think we should be using `{"revision_date" (t/plus- (t/now) (t/days 1))}` or maybe just add one additional test for every query parameter that makes sure underscores are supported.

File: system-int-test/test/cmr/system_int_test/ingest/deleted_granules_index_test.clj [hunk 12]
```diff
@@ -53,3 +71,110 @@
...
+
+(deftest deleted-granule-parameter-validation
+  (are3 [errors params]
+    (let [response (search/get-search-failure-data
+                    (search/find-deleted-granules params))]
+      (is (= (:status response) 400))
+      (is (= errors (:errors response))))
+
+    "Revision date range validation"
+    '("Revision date must be withing one year of today.")
```
Comment (chris-durbin) [inline system-int-test/test/cmr/system_int_test/ingest/deleted_granules_index_test.clj]: I don't think you need to use literal lists. `["Revision date must be withing one year of today."]` should work or just have a single error string passed into each test and change line 159 to `(is (= [error] (:errors response))))`

File: search-app/src/cmr/search/services/parameters/parameter_validation.clj (hunk 5 above)
Comment (daniel-zamora) [reply search-app/src/cmr/search/services/parameters/parameter_validation.clj]: It seems to be added automatically to the params at some point, deleted-collections has this also even though it only accepts xml

File: search-app/src/cmr/search/services/parameters/parameter_validation.clj (hunk 4 above)
Comment (daniel-zamora) [reply search-app/src/cmr/search/services/parameters/parameter_validation.clj]: This works the same as deleted collections does, I can swap the order of sanitize and validation to get it to work your way though.

File: search-app/src/cmr/search/services/parameters/parameter_validation.clj (hunk 6 above)
Comment (daniel-zamora) [reply search-app/src/cmr/search/services/parameters/parameter_validation.clj]: see other comment, this works the same as deleted collections does.  I was trying to be consistent.

File: search-app/src/cmr/search/services/query_service.clj (hunk 9 above)
Comment (daniel-zamora) [reply search-app/src/cmr/search/services/query_service.clj]: need this for esq/match-all in the filter.

File: search-app/src/cmr/search/services/query_service.clj (hunk 10 above)
Comment (daniel-zamora) [reply search-app/src/cmr/search/services/query_service.clj]: The existing query building framework requires a concept-type. There is no concept type for this query.

File: search-app/docs/api.md [hunk 13]
```diff
@@ -2640,7 +2641,7 @@ To support metadata harvesting, a harvesting client can search CMR for collectio
 
 The following search will return the last revision of the collections that are deleted since 01/20/2017.
 
-    curl -i "%CMR-ENDPOINT%/deleted-collections?revision_date=2017-01-20T00:00:00Z&pretty=true"
+    curl -i "%CMR-ENDPOINT%/deleted-granules?revision_date=2017-01-20T00:00:00Z&pretty=true"
```
Comment (chris-durbin) [inline search-app/docs/api.md]: Undo this change.

File: search-app/docs/api.md [hunk 14]
```diff
@@ -2686,6 +2687,42 @@ Tags have the following fields:
...
+Access-Control-Allow-Origin: *
+CMR-Hits: 4
+CMR-Request-Id: 03da8f3d-57b3-4ce8-bbc0-29970a4a8b30
+Content-Length: 653
+Server: Jetty(9.2.10.v20150310)
+
+[{"granule-ur":["ur2"],"revision-date":["2017-09-07T18:51:39+0000"],"parent-collection-id":["C1200000009-PROV1"],"concept-id":["G2-PROV1"],"provider-id":["PROV1"]},{"granule-ur":["ur1"],"revision-date":["2017-09-07T18:51:39+0000"],"parent-collection-id":["C1200000009-PROV1"],"concept-id":["G1-PROV1"],"provider-id":["PROV1"]},{"granule-ur":["ur3"],"revision-date":["2017-09-07T18:51:39+0000"],"parent-collection-id":["C1200000010-PROV1"],"concept-id":["G3-PROV1"],"provider-id":["PROV1"]},{"granule-ur":["ur4"],"revision-date":["2017-09-07T18:51:39+0000"],"parent-collection-id":["C1200000011-PROV2"],"concept-id":["G4-PROV2"],"provider-id":["PROV2"]}]
+```
+
+### <a name="tagging"></a> Tagging
```
Comment (chris-durbin) [inline search-app/docs/api.md]: Do you know why this tagging section is showing up as new with your PR?
